apps/ios/Shared/Model/BGManager.swift

```diff
@@ -34,35 +39,49 @@ class BGManager {
         }
     }
 
-    func schedule() {
+    func schedule(soon: Bool = false) {
         if !ChatModel.shared.ntfEnableLocal {
             logger.debug("BGManager.schedule: disabled")
             return
         }
         logger.debug("BGManager.schedule")
         let request = BGAppRefreshTaskRequest(identifier: receiveTaskId)
-        request.earliestBeginDate = Date(timeIntervalSinceNow: bgRefreshInterval)
+        request.earliestBeginDate = Date(timeIntervalSinceNow: soon ? bgRefreshInterval : runInterval)
         do {
             try BGTaskScheduler.shared.submit(request)
         } catch {
             logger.error("BGManager.schedule error: \(error.localizedDescription)")
         }
     }
 
+    var runInterval: TimeInterval {
+        switch ChatModel.shared.notificationMode {
+        case .instant: maxBgRefreshInterval
+        case .periodic: periodicBgRefreshInterval
+        case .off: bgRefreshInterval
+        }
+    }
+
+    var shouldRun: Bool {
+        Date.now.timeIntervalSince(chatLastBackgroundRunGroupDefault.get()) > runInterval
+    }
+
     private func handleRefresh(_ task: BGAppRefreshTask) {
         if !ChatModel.shared.ntfEnableLocal {
             logger.debug("BGManager.handleRefresh: disabled")
             return
         }
         logger.debug("BGManager.handleRefresh")
-        schedule()
-        if allowBackgroundRefresh() {
+        let shouldRun_ = shouldRun
+        if allowBackgroundRefresh() && shouldRun_ {
+            schedule()
```

Comment:
I don't fully understand logic here.. first, why not pass directly `interval: TimeInterval?` instead of `soon` to `schedule`? It seems as unnecessary indirection as in the else below `soon` itself depends on `runInterval`.. Second, why not always pass it to `schedule`? In if it would be `schedule(runInverval)`, in else - `schedule(shouldRun ? bgRefreshInterval : runInterval)`.

schedule could actually be even moved outside of case it seems, as it's being called in both cases now. interval parameter would be:

| allowBackgroundRefresh() && shouldRun -> runInterval (also, why not bgRefreshInterval?)
| shouldRun -> bgRefreshInterval
| otherwise -> runInterval